docs/05-rfcs/2023-01-11-sdk-spec.md

```diff
@@ -0,0 +1,1070 @@
+  log.info("File uploaded: " + file);
+});
+```
+
+The argument to an `on_<event>` method is called an event handler.
+An event handler is any resource that implements the `EventHandler<T>` interface, which contains a single method named `handle` that accepts an event of type `T`:
+
+```ts
+// wing
+
+interface EventHandler<T> {
+  inflight handle(event: T);
+}
+```
+
+<!--
+We could also decide the should be named `handle_event` or `invoke` or `run` or `call` or something else.
+-->
+
+An inflight closure that accepts a single argument of type `T` is also considered an event handler, since the compiler will automatically convert it into a resource that implements the `EventHandler<T>` interface.
+For example, the definitions of `x1` and `x2` are equivalent:
+
+```ts
+// wing
+let x1 = inflight (event: MyEvent) => {
+  log.info("Event received: " + event);
+};
+
+resource X implements EventHandler<MyEvent> {
+  inflight handle(event: MyEvent) {
+    log.info("Event received: " + event);
+  }
+}
+let x2 = new X();
+```
+
+> Reqtag: `sdk:event-handler-conversion`
+<span id="sdk:event-handler-conversion"/>
+
+In the example with `bucket.on_upload` from earlier this section, when the method is called, cloud infrastructure is automatically added to invoke the event handler's `handle` method whenever an object is added to the bucket.
+
+Users can define their own event-based APIs by making use of the `Topic` resource.
+For example, suppose you want to extend `cloud.Counter` so that it triggers an event whenever a threshold is reached or exceeded:
+
+```ts
+// wing
+
+let counter = new MyCounter(threshold: 10);
+counter.on_threshold(inflight (event: ThresholdReachedEvent) => {
+  log.info("Threshold reached: " + event.value);
+});
+```
+
+You can implement this by defining a `CounterWithThreshold` resource with an `on_threshold` method, and use a `Topic` to publish events:
+
+```ts
+// wing
+
+struct MyCounterProps extends cloud.CounterProps {
+  threshold: num;
+}
+
+resource MyCounter extends cloud.Counter {
+  _threshold: num;
+  _topic: cloud.Topic;
+
+  init(props: CounterWithThresholdProps) {
+    super(props);
+    this._threshold = props.threshold;
+    this._topic = new cloud.Topic();
+  }
+
+  inflight inc(amount?: num) {
+    let new_value = super.inc(amount);
+    if new_value >= this._threshold {
+      this._topic.publish(ThresholdReachedEvent {
+        value: new_value,
+        time_reached: time.now(),
+      });
+    }
+    return value;
+  }
+
+  // or: on_threshold(handler: EventHandler<ThresholdReachedEvent>)
+  on_threshold(handler: inflight (event: ThresholdReachedEvent) => void): void {
+    return this._topic.on_event(handler);
+  }
+}
+
+struct ThresholdReachedEvent {
+  value: num;
+  time_reached: Time;
+}
+```
+
+## Planned resources
+
+* Bucket (P1) - object storage, similar to AWS S3, Azure Blob Storage, GCP Storage
+* Queue (P1) - a message queue, similar to AWS SQS, Azure Storage Queues, GCP Pub/Sub
```

Comment:
Yes, of course. Let me remove the mention of GCP Pub/Sub from the first to avoid confusion (it just happens that I think they might both be implemented using GCP Pub/Sub).